File: t/PACSFramework/CS/MPPSWCFDemo/UI/Pages/SequencesPage.cs
// *************************************************************
// Copyright (c) 1991-2019 LEAD Technologies, Inc.              
// All Rights Reserved.                                         
// *************************************************************
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Drawing;
using System.Data;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using Leadtools.Wizard.Pages;
using MPPSWCFDemo.Broker;
using Leadtools.Wizard;
using Leadtools.Demos;
using Leadtools.Dicom.Common.DataTypes;

namespace MPPSWCFDemo.UI.Pages
{
    public partial class SequencesPage : InternalTemplatePage
    {              
        public SequencesPage()
        {
            InitializeComponent();            
            Application.Idle += new EventHandler(Application_Idle);
        }        

        void Application_Idle(object sender, EventArgs e)
        {                       
            EnableEditButtons(listViewReasonCode, buttonEditReasonCode, buttonDeleteReasonCode);
            EnableEditButtons(listViewCodeSequence, buttonEditCodeSequence, buttonDeleteCodeSequence);
            EnableEditButtons(listViewProtocol, buttonEditProtocol, buttonDeleteProtocol);
        }

        private void EnableEditButtons(ListView listView, Button edit, Button delete)
        {
            edit.Enabled = listView.SelectedItems.Count == 1;
            delete.Enabled = edit.Enabled;
        }

        protected override WizardSheet.WizardButtons GetWizardButtons()
        {
            WizardSheet.WizardButtons buttons = base.GetWizardButtons();

            buttons |= WizardSheet.WizardButtons.Option1;
            return buttons;
        }

        public override void OnOptionButton1(object sender, EventArgs e)
        {
            UpdateUI(true);
        }

        public override void OnSetActive(object sender, WizardCancelEventArgs e)
        {
            base.OnSetActive(sender, e);
            LoadSequences();
        }

        public override void OnWizardNext(object sender, WizardPageEventArgs e)
        {
            ProgressDialog dlgProgresss = new ProgressDialog();
            BrokerServiceClient client = GetWizard().Tag as BrokerServiceClient;

            base.OnWizardNext(sender, e);
            UpdateMPPS();

            dlgProgresss.Title = "Update MPPS";
            dlgProgresss.Description = "Updating..." ;
            dlgProgresss.Action = () =>
            {
                client.UpdateMPPS(MainForm.Mpps.MPPSSOPInstanceUID, MainForm.Mpps);
            };

            if (dlgProgresss.ShowDialog(this) == DialogResult.Cancel)
            {
                if (dlgProgresss.Exception != null)
                {
                    e.Cancel = true;
                    Messager.ShowError(this, dlgProgresss.Exception);
                }
            }
        }

        private void UpdateMPPS()
        {
            MainForm.Mpps.PPSDiscontinuationReasonCodeSequence.Clear();
            foreach(ListViewItem item in listViewReasonCode.Items)
            {
                PPSDiscontinuationReasonCodeSequence s = item.Tag as PPSDiscontinuationReasonCodeSequence;

                MainForm.Mpps.PPSDiscontinuationReasonCodeSequence.Add(s);
            }

            MainForm.Mpps.ProcedureCodeSequence.Clear();
            foreach (ListViewItem item in listViewCodeSequence.Items)
            {
                ProcedureCodeSequence s = item.Tag as ProcedureCodeSequence;

                MainForm.Mpps.ProcedureCodeSequence.Add(s);
            }

            MainForm.Mpps.PerformedProtocolCodeSequence.Clear();
            foreach (ListViewItem item in listViewProtocol.Items)
            {
                PerformedProtocolCodeSequence s = item.Tag as PerformedProtocolCodeSequence;

                MainForm.Mpps.PerformedProtocolCodeSequence.Add(s);
            }
        }


        /// <summary>
        /// Updates the UI.
        /// </summary>
        /// <remarks></remarks>
        private void UpdateUI(bool clear)
        {
            if (!clear)
            {
                LoadSequences();
            }
            else
            {
                ClearSequences();
            }            
        }

        private void LoadSequences()
        {
            ClearSequences();
            foreach (PPSDiscontinuationReasonCodeSequence s in MainForm.Mpps.PPSDiscontinuationReasonCodeSequence)
            {
                AddPPSDiscontinuationReasonCodeSequence(s);
            }

            foreach (ProcedureCodeSequence s in MainForm.Mpps.ProcedureCodeSequence)
            {
                AddProcedureCodeSequence(s);
            }

            foreach (PerformedProtocolCodeSequence s in MainForm.Mpps.PerformedProtocolCodeSequence)
            {
                AddPerformedProtocolCodeSequence(s);
            }
        }

        public void ClearSequences()
        {
            listViewReasonCode.Items.Clear();
            listViewCodeSequence.Items.Clear();
            listViewProtocol.Items.Clear();
        }

        private void AddPPSDiscontinuationReasonCodeSequence(PPSDiscontinuationReasonCodeSequence sequence)
        {
            ListViewItem item = listViewReasonCode.Items.Add(sequence.CodeValue);

            item.SubItems.Add(sequence.CodingSchemeDesignator);
            item.SubItems.Add(sequence.CodeMeaning);
            item.SubItems.Add(sequence.CodingSchemeVersion);
            item.SubItems.Add(sequence.OrderNumber);
            item.Tag = sequence;
        }

        private void AddProcedureCodeSequence(ProcedureCodeSequence sequence)
        {
            ListViewItem item = listViewCodeSequence.Items.Add(sequence.CodeValue);

            item.SubItems.Add(sequence.CodingSchemeDesignator);
            item.SubItems.Add(sequence.CodeMeaning);
            item.SubItems.Add(sequence.CodingSchemeVersion);
            item.SubItems.Add(sequence.OrderNumber);
            item.Tag = sequence;
        }

        private void AddPerformedProtocolCodeSequence(PerformedProtocolCodeSequence sequence)
        {
            ListViewItem item = listViewProtocol.Items.Add(sequence.CodeValue);

            item.SubItems.Add(sequence.CodingSchemeDesignator);
            item.SubItems.Add(sequence.CodeMeaning);
            item.SubItems.Add(sequence.CodingSchemeVersion);
            item.SubItems.Add(sequence.OrderNumber);
            item.Tag = sequence;
        }

        public override void OnReset()
        {
            UpdateUI(true);
        }

        private void EditSequence<T>(ListView listView, string description) where T: class
        {
            ListViewItem item = listView.SelectedItems[0];
            T sequence = item.Tag as T;
            ObjectEditor<T> editor = new ObjectEditor<T>(sequence, description);

            if (editor.ShowDialog(this) == DialogResult.OK)
            {
                Type t = typeof(T);

                item.Text = t.GetProperty("CodeValue").GetValue(editor.EditObject,null) as string;
                item.SubItems[1].Text = t.GetProperty("CodingSchemeDesignator").GetValue(editor.EditObject, null) as string;
                item.SubItems[2].Text = t.GetProperty("CodeMeaning").GetValue(editor.EditObject, null) as string;
                item.SubItems[3].Text = t.GetProperty("CodingSchemeVersion").GetValue(editor.EditObject, null) as string;
                item.SubItems[4].Text = t.GetProperty("OrderNumber").GetValue(editor.EditObject, null) as string;
                item.Tag = editor.EditObject;
            }
        }

        private void RemoveSelectedItem(ListView listview)
        {
            ListViewItem item = listview.SelectedItems[0];

            listview.Items.Remove(item);
        }

        private void buttonAddReasonCode_Click(object sender, EventArgs e)
        {
            ObjectEditor<PPSDiscontinuationReasonCodeSequence> editor = new ObjectEditor<PPSDiscontinuationReasonCodeSequence>(null, "PPS Discontinuation Reason Code Sequence");

            if (editor.ShowDialog(this) == DialogResult.OK)
            {
               PPSDiscontinuationReasonCodeSequence sequence = editor.EditObject as PPSDiscontinuationReasonCodeSequence;

               if (string.IsNullOrEmpty(sequence.CodeValue) || string.IsNullOrEmpty(sequence.CodingSchemeDesignator))
               {
                  Messager.ShowError(this, "Must provide a value for CodeValue and CodingSchemeDesignator.  Item not added to sequence.");
               }
               else
               {
                  AddPPSDiscontinuationReasonCodeSequence(editor.EditObject);
               }
            }
        }

        private void buttonEditReasonCode_Click(object sender, EventArgs e)
        {            
            EditSequence<PPSDiscontinuationReasonCodeSequence>(listViewReasonCode, "PPS Discontinuation Reason Code Sequence");
        }

        private void buttonDeleteReasonCode_Click(object sender, EventArgs e)
        {
            RemoveSelectedItem(listViewReasonCode);
        }

        private void buttonAddCodeSequence_Click(object sender, EventArgs e)
        {
            ObjectEditor<ProcedureCodeSequence> editor = new ObjectEditor<ProcedureCodeSequence>(null, "Procedure Code Sequence");

            if (editor.ShowDialog(this) == DialogResult.OK)
            {
               ProcedureCodeSequence sequence = editor.EditObject as ProcedureCodeSequence;

               if (string.IsNullOrEmpty(sequence.CodeValue) || string.IsNullOrEmpty(sequence.CodingSchemeDesignator))
               {
                  Messager.ShowError(this, "Must provide a value for CodeValue and CodingSchemeDesignator.  Item not added to sequence.");
               }
               else
               {
                  AddProcedureCodeSequence(editor.EditObject);
               }
            }
        }

        private void buttonEditCodeSequence_Click(object sender, EventArgs e)
        {            
            EditSequence<ProcedureCodeSequence>(listViewCodeSequence, "Procedure Code Sequence");
        }

        private void buttonDeleteCodeSequence_Click(object sender, EventArgs e)
        {
            RemoveSelectedItem(listViewCodeSequence);
        }

        private void buttonAddProtocol_Click(object sender, EventArgs e)
        {
            ObjectEditor<PerformedProtocolCodeSequence> editor = new ObjectEditor<PerformedProtocolCodeSequence>(null, "Performed Protocol Sequence");

            if (editor.ShowDialog(this) == DialogResult.OK)
            {
               PerformedProtocolCodeSequence sequence = editor.EditObject as PerformedProtocolCodeSequence;

               if (string.IsNullOrEmpty(sequence.CodeValue) || string.IsNullOrEmpty(sequence.CodingSchemeDesignator))
               {
                  Messager.ShowError(this, "Must provide a value for CodeValue and CodingSchemeDesignator.  Item not added to sequence.");
               }
               else
               {
                  AddPerformedProtocolCodeSequence(editor.EditObject);
               }
            }
        }        

        private void buttonEditProtocol_Click(object sender, EventArgs e)
        {            
            EditSequence<PerformedProtocolCodeSequence>(listViewProtocol, "Performed Protocol Code Sequence");
        }       

        private void buttonDeleteProtocol_Click(object sender, EventArgs e)
        {
            RemoveSelectedItem(listViewProtocol);
        }                 
    }
}

File: t/PACSFramework/VB/MPPSWCFDemo/UI/Pages/SequencesPage.vb
' *************************************************************
' Copyright (c) 1991-2019 LEAD Technologies, Inc.              
' All Rights Reserved.                                         
' *************************************************************
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Drawing
Imports System.Data
Imports System.Linq
Imports System.Text
Imports System.Windows.Forms
Imports Leadtools.Wizard.Pages
Imports MPPSWCFDemo.Broker
Imports Leadtools.Wizard
Imports Leadtools.Demos
Imports Leadtools.Dicom.Common.DataTypes

Namespace MPPSWCFDemo.UI.Pages
	Partial Public Class SequencesPage
		Inherits InternalTemplatePage
		Public Sub New()
			InitializeComponent()
			AddHandler Application.Idle, AddressOf Application_Idle
		End Sub

		Private Sub Application_Idle(ByVal sender As Object, ByVal e As EventArgs)
			EnableEditButtons(listViewReasonCode, buttonEditReasonCode, buttonDeleteReasonCode)
			EnableEditButtons(listViewCodeSequence, buttonEditCodeSequence, buttonDeleteCodeSequence)
			EnableEditButtons(listViewProtocol, buttonEditProtocol, buttonDeleteProtocol)
		End Sub

		Private Sub EnableEditButtons(ByVal listView As ListView, ByVal edit As Button, ByVal delete As Button)
			edit.Enabled = listView.SelectedItems.Count = 1
			delete.Enabled = edit.Enabled
		End Sub

		Protected Overrides Function GetWizardButtons() As WizardSheet.WizardButtons
			Dim buttons As WizardSheet.WizardButtons = MyBase.GetWizardButtons()

			buttons = buttons Or WizardSheet.WizardButtons.Option1
			Return buttons
		End Function

		Public Overrides Sub OnOptionButton1(ByVal sender As Object, ByVal e As EventArgs)
			UpdateUI(True)
		End Sub

		Public Overrides Sub OnSetActive(ByVal sender As Object, ByVal e As WizardCancelEventArgs)
			MyBase.OnSetActive(sender, e)
			LoadSequences()
		End Sub

	   Public _clientSequencesPageOnWizardNext As BrokerServiceClient
		Public Overrides Sub OnWizardNext(ByVal sender As Object, ByVal e As WizardPageEventArgs)
			Dim dlgProgresss As New ProgressDialog()
			_clientSequencesPageOnWizardNext = TryCast(GetWizard().Tag, BrokerServiceClient)

			MyBase.OnWizardNext(sender, e)
			UpdateMPPS()

			dlgProgresss.Title = "Update MPPS"
			dlgProgresss.Description = "Updating..."
			dlgProgresss.Action = AddressOf AnonymousMethod1

			If dlgProgresss.ShowDialog(Me) = DialogResult.Cancel Then
				If dlgProgresss.Exception IsNot Nothing Then
					e.Cancel = True
					Messager.ShowError(Me, dlgProgresss.Exception)
				End If
			End If
		End Sub
		Private Sub AnonymousMethod1()
			_clientSequencesPageOnWizardNext.UpdateMPPS(MainForm.Mpps.MPPSSOPInstanceUID, MainForm.Mpps)
		End Sub

		Private Sub UpdateMPPS()
			MainForm.Mpps.PPSDiscontinuationReasonCodeSequence.Clear()
			For Each item As ListViewItem In listViewReasonCode.Items
				Dim s As PPSDiscontinuationReasonCodeSequence = TryCast(item.Tag, PPSDiscontinuationReasonCodeSequence)

				MainForm.Mpps.PPSDiscontinuationReasonCodeSequence.Add(s)
			Next item

			MainForm.Mpps.ProcedureCodeSequence.Clear()
			For Each item As ListViewItem In listViewCodeSequence.Items
				Dim s As ProcedureCodeSequence = TryCast(item.Tag, ProcedureCodeSequence)

				MainForm.Mpps.ProcedureCodeSequence.Add(s)
			Next item

			MainForm.Mpps.PerformedProtocolCodeSequence.Clear()
			For Each item As ListViewItem In listViewProtocol.Items
				Dim s As PerformedProtocolCodeSequence = TryCast(item.Tag, PerformedProtocolCodeSequence)

				MainForm.Mpps.PerformedProtocolCodeSequence.Add(s)
			Next item
		End Sub


		''' <summary>
		''' Updates the UI.
		''' </summary>
		''' <remarks></remarks>
		Private Sub UpdateUI(ByVal clear As Boolean)
			If (Not clear) Then
				LoadSequences()
			Else
				ClearSequences()
			End If
		End Sub

		Private Sub LoadSequences()
			ClearSequences()
			For Each s As PPSDiscontinuationReasonCodeSequence In MainForm.Mpps.PPSDiscontinuationReasonCodeSequence
				AddPPSDiscontinuationReasonCodeSequence(s)
			Next s

			For Each s As ProcedureCodeSequence In MainForm.Mpps.ProcedureCodeSequence
				AddProcedureCodeSequence(s)
			Next s

			For Each s As PerformedProtocolCodeSequence In MainForm.Mpps.PerformedProtocolCodeSequence
				AddPerformedProtocolCodeSequence(s)
			Next s
		End Sub

		Public Sub ClearSequences()
			listViewReasonCode.Items.Clear()
			listViewCodeSequence.Items.Clear()
			listViewProtocol.Items.Clear()
		End Sub

		Private Sub AddPPSDiscontinuationReasonCodeSequence(ByVal sequence As PPSDiscontinuationReasonCodeSequence)
			Dim item As ListViewItem = listViewReasonCode.Items.Add(sequence.CodeValue)

			item.SubItems.Add(sequence.CodingSchemeDesignator)
			item.SubItems.Add(sequence.CodeMeaning)
			item.SubItems.Add(sequence.CodingSchemeVersion)
			item.SubItems.Add(sequence.OrderNumber)
			item.Tag = sequence
		End Sub

		Private Sub AddProcedureCodeSequence(ByVal sequence As ProcedureCodeSequence)
			Dim item As ListViewItem = listViewCodeSequence.Items.Add(sequence.CodeValue)

			item.SubItems.Add(sequence.CodingSchemeDesignator)
			item.SubItems.Add(sequence.CodeMeaning)
			item.SubItems.Add(sequence.CodingSchemeVersion)
			item.SubItems.Add(sequence.OrderNumber)
			item.Tag = sequence
		End Sub

		Private Sub AddPerformedProtocolCodeSequence(ByVal sequence As PerformedProtocolCodeSequence)
			Dim item As ListViewItem = listViewProtocol.Items.Add(sequence.CodeValue)

			item.SubItems.Add(sequence.CodingSchemeDesignator)
			item.SubItems.Add(sequence.CodeMeaning)
			item.SubItems.Add(sequence.CodingSchemeVersion)
			item.SubItems.Add(sequence.OrderNumber)
			item.Tag = sequence
		End Sub

		Public Overrides Sub OnReset()
			UpdateUI(True)
		End Sub

		Private Sub EditSequence(Of T As Class)(ByVal listView As ListView, ByVal description As String)
			Dim item As ListViewItem = listView.SelectedItems(0)
			Dim sequence As T = TryCast(item.Tag, T)
			Dim editor As New ObjectEditor(Of T)(sequence, description)

			If editor.ShowDialog(Me) = DialogResult.OK Then
				Dim myType As Type = GetType(T)

				item.Text = TryCast(myType.GetProperty("CodeValue").GetValue(editor.EditObject,Nothing), String)
				item.SubItems(1).Text = TryCast(myType.GetProperty("CodingSchemeDesignator").GetValue(editor.EditObject, Nothing), String)
				item.SubItems(2).Text = TryCast(myType.GetProperty("CodeMeaning").GetValue(editor.EditObject, Nothing), String)
				item.SubItems(3).Text = TryCast(myType.GetProperty("CodingSchemeVersion").GetValue(editor.EditObject, Nothing), String)
				item.SubItems(4).Text = TryCast(myType.GetProperty("OrderNumber").GetValue(editor.EditObject, Nothing), String)
				item.Tag = editor.EditObject
			End If
		End Sub

		Private Sub RemoveSelectedItem(ByVal listview As ListView)
			Dim item As ListViewItem = listview.SelectedItems(0)

			listview.Items.Remove(item)
		End Sub

		Private Sub buttonAddReasonCode_Click(ByVal sender As Object, ByVal e As EventArgs) Handles buttonAddReasonCode.Click
			Dim editor As New ObjectEditor(Of PPSDiscontinuationReasonCodeSequence)(Nothing, "PPS Discontinuation Reason Code Sequence")

			If editor.ShowDialog(Me) = DialogResult.OK Then
			   Dim sequence As PPSDiscontinuationReasonCodeSequence = TryCast(editor.EditObject, PPSDiscontinuationReasonCodeSequence)

			   If String.IsNullOrEmpty(sequence.CodeValue) OrElse String.IsNullOrEmpty(sequence.CodingSchemeDesignator) Then
				  Messager.ShowError(Me, "Must provide a value for CodeValue and CodingSchemeDesignator.  Item not added to sequence.")
			   Else
				  AddPPSDiscontinuationReasonCodeSequence(editor.EditObject)
			   End If
			End If
		End Sub

		Private Sub buttonEditReasonCode_Click(ByVal sender As Object, ByVal e As EventArgs) Handles buttonEditReasonCode.Click
			EditSequence(Of PPSDiscontinuationReasonCodeSequence)(listViewReasonCode, "PPS Discontinuation Reason Code Sequence")
		End Sub

		Private Sub buttonDeleteReasonCode_Click(ByVal sender As Object, ByVal e As EventArgs) Handles buttonDeleteReasonCode.Click
			RemoveSelectedItem(listViewReasonCode)
		End Sub

		Private Sub buttonAddCodeSequence_Click(ByVal sender As Object, ByVal e As EventArgs) Handles buttonAddCodeSequence.Click
			Dim editor As New ObjectEditor(Of ProcedureCodeSequence)(Nothing, "Procedure Code Sequence")

			If editor.ShowDialog(Me) = DialogResult.OK Then
			   Dim sequence As ProcedureCodeSequence = TryCast(editor.EditObject, ProcedureCodeSequence)

			   If String.IsNullOrEmpty(sequence.CodeValue) OrElse String.IsNullOrEmpty(sequence.CodingSchemeDesignator) Then
				  Messager.ShowError(Me, "Must provide a value for CodeValue and CodingSchemeDesignator.  Item not added to sequence.")
			   Else
				  AddProcedureCodeSequence(editor.EditObject)
			   End If
			End If
		End Sub

		Private Sub buttonEditCodeSequence_Click(ByVal sender As Object, ByVal e As EventArgs) Handles buttonEditCodeSequence.Click
			EditSequence(Of ProcedureCodeSequence)(listViewCodeSequence, "Procedure Code Sequence")
		End Sub

		Private Sub buttonDeleteCodeSequence_Click(ByVal sender As Object, ByVal e As EventArgs) Handles buttonDeleteCodeSequence.Click
			RemoveSelectedItem(listViewCodeSequence)
		End Sub

		Private Sub buttonAddProtocol_Click(ByVal sender As Object, ByVal e As EventArgs) Handles buttonAddProtocol.Click
			Dim editor As New ObjectEditor(Of PerformedProtocolCodeSequence)(Nothing, "Performed Protocol Sequence")

			If editor.ShowDialog(Me) = DialogResult.OK Then
			   Dim sequence As PerformedProtocolCodeSequence = TryCast(editor.EditObject, PerformedProtocolCodeSequence)

			   If String.IsNullOrEmpty(sequence.CodeValue) OrElse String.IsNullOrEmpty(sequence.CodingSchemeDesignator) Then
				  Messager.ShowError(Me, "Must provide a value for CodeValue and CodingSchemeDesignator.  Item not added to sequence.")
			   Else
				  AddPerformedProtocolCodeSequence(editor.EditObject)
			   End If
			End If
		End Sub

		Private Sub buttonEditProtocol_Click(ByVal sender As Object, ByVal e As EventArgs) Handles buttonEditProtocol.Click
			EditSequence(Of PerformedProtocolCodeSequence)(listViewProtocol, "Performed Protocol Code Sequence")
		End Sub

		Private Sub buttonDeleteProtocol_Click(ByVal sender As Object, ByVal e As EventArgs) Handles buttonDeleteProtocol.Click
			RemoveSelectedItem(listViewProtocol)
		End Sub
	End Class
End Namespace
